api/insight/apimiddleware.go

```diff
@@ -91,3 +109,161 @@ func (c *insightApiContext) PostBroadcastTxCtx(next http.Handler) http.Handler {
 		next.ServeHTTP(w, r.WithContext(ctx))
 	})
 }
+
+// GetAddressCtx retrieves the ctxAddress data from the request context. If not
+// set, the return value is an empty string.
+func (c *insightApiContext) GetAddressCtx(r *http.Request) string {
+	address, ok := r.Context().Value(ctxAddress).(string)
+	if !ok {
+		apiLog.Trace("address not set")
+		return ""
+	}
+	return address
+}
+
+// GetToCtx retrieves the ctxTo data ("to") from the request context. If not
+// set, the return value ok is false.
+func (c *insightApiContext) GetToCtx(r *http.Request) (int64, bool) {
+	to, ok := r.Context().Value(ctxTo).(int)
+	if !ok {
+		apiLog.Errorf("to param not set")
+		return int64(0), false
+	}
+	return int64(to), true
+}
+
+// GetFromCtx retrieves the ctxFrom data ("from") from the request context.
+// If not set, the return value ok is false.
+func (c *insightApiContext) GetFromCtx(r *http.Request) (int64, bool) {
+	from, ok := r.Context().Value(ctxFrom).(int)
+	if !ok {
+		apiLog.Errorf("from param not set")
+		return int64(0), false
+	}
+	return int64(from), true
+}
+
+// GetNoAsmCtx retrieves the ctxNoAsm data ("noAsm") from the request context.
+// If not set, the return value is false.
+func (c *insightApiContext) GetNoAsmCtx(r *http.Request) bool {
+	noAsm, ok := r.Context().Value(ctxNoAsm).(bool)
+	if !ok {
+		return false
+	}
+	return noAsm
+}
+
+// GetNoScriptSigCtx retrieves the ctxNoScriptSig data ("noScriptSig") from the
+// request context. If not set, the return value is false.
+func (c *insightApiContext) GetNoScriptSigCtx(r *http.Request) bool {
+	noScriptSig, ok := r.Context().Value(ctxNoScriptSig).(bool)
+	if !ok {
+		return false
+	}
+	return noScriptSig
+}
+
+// GetNoSpentCtx retrieves the ctxNoSpent data ("noSpent") from the
+// request context. If not set, the return value is false.
+func (c *insightApiContext) GetNoSpentCtx(r *http.Request) bool {
+	noNoSpent, ok := r.Context().Value(ctxNoSpent).(bool)
+	if !ok {
+		return false
+	}
+	return noNoSpent
+}
+
+// PaginationCtx will parse the query parameters for from/to values.
+func (c *insightApiContext) PaginationCtx(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		ctx := r.Context()
+		to, from := r.FormValue("to"), r.FormValue("from")
+		fromint, err := strconv.Atoi(from)
+		if err == nil {
+			ctx = context.WithValue(r.Context(), ctxFrom, fromint)
+		}
+		toint, err := strconv.Atoi(to)
+		if err == nil {
+			ctx = context.WithValue(ctx, ctxTo, toint)
+		}
+		next.ServeHTTP(w, r.WithContext(ctx))
+	})
+}
+
+// Process params given in post body for an addrs endpoint
+func (c *insightApiContext) PostAddrsTxsCtx(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		var req apitypes.InsightMultiAddrsTx
+		var from, to, noAsm, noScriptSig, noSpent int64
+		body, err := ioutil.ReadAll(r.Body)
```

Comment:
Usually we'd first check the request's method (ensure POST), the Content-Type, and any other headers we expect to get.  For example, here's a little file upload/download web app I put together recently:

https://github.com/chappjc/webfiles/blob/c5e702e2b6311a340722f637b8f98933fc63c8ba/server/server.go#L206

Reply:
I just copied the way I saw dcrd do it which I don't think checks that, but I'll add in content checking as an error response along those lines would have saved me hours when I was trying to figure out how insight post params work.

Reply:
This example is file upload, so it's almost certainly a different content type, but by checking before reading the body we'll potentially avoid transferring useless data

Reply:
chi mux has middleware to check content type so I added that in.